v2/cmd/controller/app/setup.go

```diff
@@ -156,39 +155,44 @@ func SetupControllerManager(ctx context.Context, setupLog logr.Logger, flgs Flag
 		setupLog.Error(err, "failed to initialize CRD client")
 		os.Exit(1)
 	}
-
-	goalCRDs, err := crdManager.LoadOperatorCRDs(crdmanagement.CRDLocation, cfg.PodNamespace)
-	if err != nil {
-		setupLog.Error(err, "failed to load CRDs from disk")
-		os.Exit(1)
-	}
-
 	existingCRDs, err := crdManager.ListOperatorCRDs(ctx)
 	if err != nil {
 		setupLog.Error(err, "failed to list current CRDs")
 		os.Exit(1)
 	}
 
-	// We only apply CRDs if we're in webhooks mode. No other mode will have CRD CRUD permissions
-	if cfg.OperatorMode.IncludesWebhooks() {
-		var installationInstructions []*crdmanagement.CRDInstallationInstruction
-		installationInstructions, err = crdManager.DetermineCRDsToInstallOrUpgrade(goalCRDs, existingCRDs, flgs.CRDPatterns)
+	// By default, assume the existing CRDs are the goal CRDs. If CRD management is enabled, we will
+	// load the goal CRDs from disk and apply them.
+	goalCRDs := existingCRDs
+	if flgs.EnableCRDManagement {
```

Comment:
This will lead to some very hard to diagnose failure modes.

The CRD versions in the cluster **must exactly** match the version of ASO that's being run. Upgrading either one in isolation will create an environment where some, if not all resources, will not work.

The underlying issue is to do with which version of each resource is tagged as the storage version - if an ASO release introduces a new version of a resource, the storage version changes (e.g. from `v1api20200401storage` to `v120230801storage`); a mismatch of versions between the CRDs and the running ASO operator is a fatal error.

I think we need to retain the check, but provide an option on what to do if a CRD mismatch is discovered. I can think of three options:

* auto-upgrade - the existing behaviour
* abort - the operator refuses to run, with suitable logging to inform users
* degrade - the operator runs, but only supporting correctly versioned resources

Reply:
We considered using a central "hub" storage version during design, but that had [some](https://azure.github.io/azure-service-operator/design/versioning/#alternative-fixed-storage-version) [problems](https://azure.github.io/azure-service-operator/design/versioning/case-studies/fixed-storage-version/).

The approach we've taken is [chained storage versions](https://azure.github.io/azure-service-operator/design/versioning/case-studies/chained-storage-versions/); that case study goes into a great deal of information.

Reply:
After discussions on Slack, I better understand the usefulness of this flag. 

Perhaps we rename the flag to better indicate that the user is _disabling_ standard functionality, rather than _enabling_ optional functionality?

For example, instead of `--enable-crd-management` with values `true` (the default) and `false`, what if it was `--crd-compatibility-check` with values of `auto-upgrade` (the default), `verify` and `none` ?

Reply:
I agree with the reversed phrasing, (about disabling a feature versus enabling one), if we were leaving this a boolean. I'm not sold on the `crd-compatibility-check` naming, as IMO "upgrading the CRDs" isn't really a compatibility check and that's the default mode.

Maybe `crd-management-mode` `auto|none`, which the default being `auto`? I also agree that a `verify` might make sense but not sure we need to add it in this PR.

Reply:
I like `crd-management-mode` with the options of `auto|none` - if we were to add `verify` in a future PR, that wouldn't be a breaking change.